src/highlight/HighlightManager.ts

```diff
@@ -0,0 +1,195 @@
+    return [];
+  }
+
+  public GetDecorations(fileName: string): DecorationOptions[] {
+    const idx = this.highlightCollection.findIndex(hc => hc.fileName === fileName);
+    if (idx > -1) {
+      return this.highlightCollection[idx].highlights.map<DecorationOptions>(h => {
+        return {
+          hoverMessage: `From ${h.userName === 'self'? 'You' : h.userName} ${h.comments !== undefined ? h.comments : ''}`,
+          range: h.range
+        };
+      });
+    }
+    return [];
+  }
+
+  public Add(document: TextDocument, userName: string, startLine: number, endLine?: number, comments?: string): void {
+    if (!endLine) {
+      endLine = startLine;
+    }
+
+    const range = new Range(
+      new Position(--startLine, 0),
+      new Position(--endLine, document.lineAt(endLine).text.length)
+    );
+
+    const highlight = new Highlight(userName, range, comments);
+
+    const idx = this.highlightCollection.findIndex(h => h.fileName === document.fileName);
+    if (idx > -1) {
+      if (!this.highlightCollection[idx].highlights.some(h => (h.userName === userName || userName === 'self') && h.startLine <= startLine && h.endLine >= endLine!)) {
+        this.highlightCollection[idx].highlights.push(highlight);
+      }
+    }
+    else {
+      this.highlightCollection.push({
+        fileName: document.fileName,
+        highlights: [highlight]
+      });
+    }
+
+    this._onHighlightsChanged.fire();
+  }
+
+  public Remove(document: TextDocument, userName: string, lineNumber: number, deferRefresh?: boolean): void;
+  public Remove(fileName: string, userName: string, lineNumber: number, deferRefresh?: boolean): void;
+  public Remove(documentOrFileName: TextDocument | string, userName: string, lineNumber: number, deferRefresh: boolean = false): void {
+    if (!isString(documentOrFileName)) {
+      documentOrFileName = documentOrFileName.fileName;
+    }
+
+    const idx = this.highlightCollection.findIndex(h => h.fileName === documentOrFileName);
+    if (idx > -1) {
+      const hidx = this.highlightCollection[idx].highlights.findIndex(h => (h.userName === userName || userName === 'self') && h.startLine <= lineNumber && h.endLine >= lineNumber);
+      if (hidx > -1) {
+        this.highlightCollection[idx].highlights.splice(hidx, 1);
+      }
+      if (!deferRefresh) {
+        this._onHighlightsChanged.fire();
+      }
+    }
+  }
+
+  public Refresh() {
+    this._onHighlightsChanged.fire();
+  }
+
+  public Clear(service?: string) {
+    if (service) {
+      this.highlightCollection.forEach(hc => {
+        const highlightsToRemove = hc.highlights.filter(h => h.userName.indexOf(`${service}:`) > -1);
+        highlightsToRemove.forEach(h => {
+          this.Remove(hc.fileName, h.userName, h.startLine, true);
+        });
+      });
+    }
+    else {
+      this.highlightCollection = new Array<HighlightCollection>();
+    }
+    this._onHighlightsChanged.fire();
+  }
+
+  public Rename(oldName: string, newName: string) {
```

Comment:
Oh wow hadn't thought about that happening. Good call 👍

Reply:
Yup! I ran into that when testing highlights in an untitled (e.g. new) file and then saving that file (giving it a new name).